src/Storages/NATS/NATSProducer.cpp

```diff
@@ -117,11 +119,15 @@ void NATSProducer::stopProducingTask()
 
 void NATSProducer::startProducingTaskLoop()
 {
+    SCOPE_EXIT(nats_ReleaseThreadMemory());
+
     try
     {
         while (!payloads.isFinishedAndEmpty())
         {
-            if (!connection->isConnected())
+            if (shutdown_called)
+                throw Exception(ErrorCodes::UNFINISHED, "Operation aborted");
```

Comment:
I think we should publish the messages from payloads, and only prevent new messages to be pushed to payloads in case the table is shutting down. We cannot guarantee to write all the messages (there can be network issues or any other issues), but we should at least try our best to push all the messages.